### PR TITLE
[coreimage] Update for Xcode 11[.1]

### PR DESCRIPTION
**Pull request by spouliot**
```
Incomplete (draft) - it's missing a bunch of `[Obsolete]` but those are
not breaking changes so the _core_ can be reviewed.

Pay attention to naming (many conflicts) and comment if you have better
names suggestions.

It's also a big PR with a lot of copy/paste so watch out for typos!
```

Apple decided to expose most (but not all) `CIFilter` using protocols
(instead of weakly named dictionaries). Most of this maps well with
our strong bindings but there are cases where we:

* missing some properties (easy, there were added); or
* used a different types [1] and new members were required

[1] Often ours are better (using `float` for a `bool` is not optimal)
but we do not have `[BindAs]` for protocols :( to _fix_ them

### CODE REVIEW

File: src/coreimage.cs [hunk 1]
```diff
@@ -2768,22 +2916,28 @@ interface CICodeGenerator {
...
 	[Mac (10,10)]
 	[BaseType (typeof (CICodeGenerator))]
-	interface CIAztecCodeGenerator {
+	interface CIAztecCodeGenerator : ICIAztecCodeGeneratorProtocol {
 
 		[CoreImageFilterProperty ("inputCompactStyle")]
 		bool CompactStyle { get; set; }
 
+		[CoreImageFilterProperty ("inputCompactStyle")]
+		float FCompactStyle { get; set; }
```
Comment (rolfbjarne) [inline src/coreimage.cs]: What about `InputCompactStyle`?

If not, then I think I prefer `FloatCompactStyle`, it's just as ugly as `FCompactStyle`, but more informative (and the same goes for all the other new float properties).

File: src/coreimage.cs [hunk 2]
```diff
@@ -4646,10 +5006,13 @@ interface CIFaceBalance {
 
 	[iOS (9,3)]
 	[TV (9,2)]
-	[Availability (Introduced = Platform.Mac_10_10, Obsoleted = Platform.Mac_10_11)]  // FIXME: Is htis actually deprecated?  Seems to be missing in El Capitan
+//	[Availability (Introduced = Platform.Mac_10_10, Obsoleted = Platform.Mac_10_11)]  // FIXME: Is htis actually deprecated?  Seems to be missing in El Capitan
```
[the remark this answers is not in the document]
Comment (spouliot) [reply src/coreimage.cs]: Testing, it seems El Capitan was an exception (where it was not working) and not a real/undocumented obsolete filter. I'll look at (old macOS) bot results

File: src/coreimage.cs (hunk 1 above)
Comment (spouliot) [reply src/coreimage.cs]: yeah, ideally there's a common naming fix that can be used - at least for the `F` and `N` prefixes. I'll check if `Input` works well on all of them :)

File: tests/introspection/ApiCoreImageFiltersTest.cs [hunk 3]
```diff
@@ -76,7 +77,7 @@ protected virtual bool Skip (string nativeName)
 			// uncomment calls to `GenerateBinding` to use introspection code to generate the skeleton binding code and complete it
 			// e.g. picking better types like `bool` instead of `NSNumber'
 			default:
- 				return false;
+				return false;
```
Comment (VincentDondain) [inline tests/introspection/ApiCoreImageFiltersTest.cs]: Noise

File: tests/introspection/ApiCoreImageFiltersTest.cs [hunk 4]
```diff
@@ -120,7 +121,7 @@ public void CheckManagedFilters ()
 
 				// we need to skip the filters that are not supported by the executing version of iOS
 				if (Skip (t))
-					continue; 
+					continue;
```
Comment (VincentDondain) [inline tests/introspection/ApiCoreImageFiltersTest.cs]: Noise

File: src/coreimage.cs [hunk 5]
```diff
@@ -1346,6 +1382,16 @@ interface CIImage : NSSecureCoding, NSCopying {
 		[Wrap ("FromCGImage (image, options == null ? null : options.Dictionary)")]
 		CIImage FromCGImage (CGImage image, [NullAllowed] CIImageInitializationOptionsWithMetadata options);
 
+		[iOS (13,0)][TV (13,0)][Watch (6,0)][Mac (10,15)]
+		[Static]
+		[Export ("imageWithCGImageSource:index:options:")]
```
Comment (VincentDondain) [inline src/coreimage.cs]:
```suggestion
		[EditorBrowsable (EditorBrowsableState.Advanced)]
		[Export ("imageWithCGImageSource:index:options:")]
```

I think you want to encourage people to use the strongly typed version.

File: src/coreimage.cs [hunk 6]
```diff
@@ -2645,10 +2788,15 @@ interface CIImageProcessorKernel {
 	[iOS (8,0)]
 	[Mac (10,10)]
 	[BaseType (typeof (CIFilter))]
-	interface CIAccordionFoldTransition {
+	interface CIAccordionFoldTransition : ICIAccordionFoldTransitionProtocol {
 
+#if !XAMCORE_4_0
+		[Obsolete ("")]
```
Comment (VincentDondain) [inline src/coreimage.cs]:
```suggestion
		[Obsolete ("Use 'FoldCount' instead.")]
```

File: src/coreimage.cs [hunk 7]
```diff
@@ -3385,11 +3580,15 @@ interface CIFalseColor {
 
 	[CoreImageFilter]
 	[BaseType (typeof (CITransitionFilter))]
-	interface CIFlashTransition {
+	interface CIFlashTransition : ICIFlashTransitionProtocol {
 
+#if !XAMCORE_4_0
 		// for some reason we prefixed all Striation* with Max - API compatibility
 		[CoreImageFilterProperty ("inputStriationContrast")]
```
Comment (VincentDondain) [inline src/coreimage.cs]:
```suggestion
		[Obsolete ("Use 'StriationContrast' instead.")]
		[CoreImageFilterProperty ("inputStriationContrast")]
```

File: src/coreimage.cs [hunk 8]
```diff
@@ -5735,23 +6150,3278 @@ interface CIPerspectiveRotate {
...
+	[TV (13,0)]
+	[Mac (10,15)]
+	[Protocol (Name = "CIColorControls")]
+	interface CIColorControlsProtocol : CIFilterProtocol {
+
+		[Abstract]
+		[NullAllowed, Export ("inputImage", ArgumentSemantic.Retain)]
+		CIImage InputImage { get; set; }
+
+		// @required @property (nonatomic) float saturation;
```
Comment (VincentDondain) [inline src/coreimage.cs]: Forgot to remove?

File: src/coreimage.cs (hunk 7 above)
Comment (spouliot) [reply src/coreimage.cs]: noise ;-)

File: src/coreimage.cs (hunk 6 above)
Comment (spouliot) [reply src/coreimage.cs]: yup, like mentioned in the PR message, obsolete are not complete in the draft :)

File: src/coreimage.cs (hunk 7 above)
Comment (VincentDondain) [reply src/coreimage.cs]: Yea, read your description and forgot about it after 10k lines reviewed :P

File: src/coreimage.cs [hunk 9]
```diff
@@ -4812,11 +4940,14 @@ interface CITriangleKaleidoscope : ICITriangleKaleidoscopeProtocol {
 		[CoreImageFilterProperty ("inputSize")]
 		float Size { get; set; }
 
+#if !XAMCORE_4_0
+		[Obsolete ("Use 'InputPoint0' instead.")]
```
Comment (VincentDondain) [inline src/coreimage.cs]: Shouldn't this be pointing to `InputPoint`? Or `CGPoint InputPoint { get; set; }` below should be `InputPoint0`?

**Comment by spouliot** [comment]
Most (12/13) failures are (or will be) fixed by https://github.com/xamarin/xamarin-macios/pull/7154
or a back port of it to the xcode11.1 branch

**Comment by spouliot** [comment]
Closing in favour or updated/rebased https://github.com/xamarin/xamarin-macios/pull/7216

**Pull request closed without merging**